kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/DeltaTableWritesSuite.scala

```diff
@@ -1321,6 +1321,369 @@ class DeltaTableWritesSuite extends DeltaTableWriteSuiteBase with ParquetSuiteBa
+          .withSchema(engine, newSchema)
+          .build(engine).commit(engine, emptyIterable())
+      }
+
+      assert(e.getMessage.contains(
+        "Map field c must have exactly 2 nested IDs"))
+    }
+  }
+
+  test("Test fail updating schema if map columns are missing nested ids") {
+    withTempDirAndEngine { (tablePath, engine) =>
+      val table = Table.forPath(engine, tablePath)
+      val schema = new StructType()
+        .add("a", IntegerType.INTEGER, true)
+
+      createTxn(engine, tablePath, isNewTable = true, schema, partCols = Seq.empty,
+        tableProperties = Map(TableConfig.COLUMN_MAPPING_MODE.getKey -> "id",
+          TableConfig.ICEBERG_COMPAT_V2_ENABLED.getKey -> "true")).commit(engine, emptyIterable())
+
+      val currentSchema = table.getLatestSnapshot(engine).getSchema
+      val newSchema = new StructType()
+        .add("a", StringType.STRING, true, currentSchema.get("a").getMetadata)
+        .add("c", new MapType(
+          StringType.STRING, StringType.STRING, false), true,
+          FieldMetadata.builder().putLong(ColumnMapping.COLUMN_MAPPING_ID_KEY, 2)
+            .putString(ColumnMapping.COLUMN_MAPPING_PHYSICAL_NAME_KEY, "c").build())
+
+      val e = intercept[IllegalArgumentException] {
+        table.createTransactionBuilder(engine, testEngineInfo, Operation.MANUAL_UPDATE)
+          .withSchema(engine, newSchema)
+          .build(engine).commit(engine, emptyIterable())
+      }
+
+      assert(e.getMessage.contains(
+        "Map field c must have exactly 2 nested IDs"))
+    }
+  }
+
+  test("Test fail updating schema if array columns is missing nested id") {
+    withTempDirAndEngine { (tablePath, engine) =>
+      val table = Table.forPath(engine, tablePath)
+      val schema = new StructType()
+        .add("a", IntegerType.INTEGER, true)
+
+      createTxn(engine, tablePath, isNewTable = true, schema, partCols = Seq.empty,
+        tableProperties = Map(TableConfig.COLUMN_MAPPING_MODE.getKey -> "id",
+          TableConfig.ICEBERG_COMPAT_V2_ENABLED.getKey -> "true")).commit(engine, emptyIterable())
+
+      val currentSchema = table.getLatestSnapshot(engine).getSchema
+      val newSchema = new StructType()
+        .add("a", StringType.STRING, true, currentSchema.get("a").getMetadata)
+        .add("c", new ArrayType(StringType.STRING, false), true,
+          FieldMetadata.builder().putLong(ColumnMapping.COLUMN_MAPPING_ID_KEY, 2)
+            .putString(ColumnMapping.COLUMN_MAPPING_PHYSICAL_NAME_KEY, "c").build())
+
+      val e = intercept[IllegalArgumentException] {
+        table.createTransactionBuilder(engine, testEngineInfo, Operation.MANUAL_UPDATE)
+          .withSchema(engine, newSchema)
+          .build(engine).commit(engine, emptyIterable())
+      }
+
+      assert(e.getMessage.contains(
+        "Array field c must have exactly 1 nested ID"))
+    }
+  }
+
+  test("Test updating schema with adding an array and map type") {
```

Comment:
So this test and the first test `Test set schema on existing table` are the only "positive" cases covered. Are there others we should cover here?

Should renaming, moving, etc. be covered here?